src/mongo/shell/shell_options.cpp

```diff
@@ -308,6 +310,17 @@ Status storeMongoShellOptions(const moe::Environment& params,
      *   - it doesn't end in '.js' and it doesn't specify a path to an existing file */
     if (params.count("dbaddress")) {
         string dbaddress = params["dbaddress"].as<string>();
+
+        /* Support Connection String URI Format */
+        if (dbaddress.find_first_of("mongodb://") != std::string::npos) {
+            std::cmatch cm;
+            std::regex_match(dbaddress.c_str(), cm, std::regex("mongodb://([^:]+?):([^@]+?)@(.*)"));
```

Comment:
FYI, this is the RE that the legacy C++ driver uses: https://github.com/mongodb/mongo-cxx-driver/blob/legacy/src/mongo/client/dbclient.cpp#L82-L97